src/hub.rs

```diff
@@ -240,40 +272,94 @@ impl Hub {
         }
     }
 
-    pub(crate) fn update_graph(&mut self) {
-        let mut cursor = self.nodes.cursor();
-        while let Some((left, mut item, _)) = cursor.next() {
-            if !item.visible {
-                item.world_visible = false;
-                continue;
-            }
-            let (visibility, affilation, transform) = match item.parent {
-                Some(ref parent_ptr) => match left.get(parent_ptr) {
-                    Some(parent) => (
-                        parent.world_visible,
-                        parent.scene_id,
-                        parent.world_transform.concat(&item.transform),
-                    ),
-                    None => {
-                        error!("Parent node was created after the child, ignoring");
-                        (false, item.scene_id, item.transform)
-                    }
-                },
-                None => (true, item.scene_id, item.transform),
-            };
-            item.world_visible = visibility;
-            item.scene_id = affilation;
-            item.world_transform = transform;
-        }
-    }
-
     pub(crate) fn update_mesh(
         &mut self,
         mesh: &DynamicMesh,
     ) {
-        match self.get_mut(&mesh).sub_node {
+        match self[mesh].sub_node {
             SubNode::Visual(_, ref mut gpu_data) => gpu_data.pending = Some(mesh.dynamic.clone()),
             _ => unreachable!(),
         }
     }
+
+    fn walk_impl(
+        &self, base: &Option<NodePointer>, only_visible: bool
+    ) -> TreeWalker {
+        let default_stack_size = 10;
+        let mut walker = TreeWalker {
+            hub: self,
+            only_visible,
+            stack: Vec::with_capacity(default_stack_size),
+        };
+        walker.descend(base);
+        walker
+    }
+
+    pub(crate) fn walk(&self, base: &Option<NodePointer>) -> TreeWalker {
```

Comment:
Nit: Perhaps `walk_visible`?

Reply:
As in froggy, it's better to use shorter names.

Reply:
it's for internal use only anyhow. Most of the time we are only concerned about visible nodes

Reply:
Yes, I remember your strategy :smile:
![alt](https://screenshots.firefoxusercontent.com/images/9194b8ff-3e7c-4ce6-ae9a-5fe0647cc19d.png)